packages/web3-providers-ipc/src/index.ts

```diff
@@ -0,0 +1,193 @@
+import { EventEmitter } from 'events';
+import { existsSync } from 'fs';
+import { Socket } from 'net';
+import {
+	ConnectionNotOpenError,
+	DeferredPromise,
+	EthExecutionAPI,
+	InvalidClientError,
+	InvalidConnectionError,
+	JsonRpcId,
+	JsonRpcNotification,
+	JsonRpcResponse,
+	JsonRpcResponseWithError,
+	JsonRpcResponseWithResult,
+	JsonRpcResult,
+	Web3APIMethod,
+	Web3APIPayload,
+	Web3APIReturnType,
+	Web3APISpec,
+	Web3BaseProvider,
+	Web3BaseProviderCallback,
+	Web3BaseProviderStatus,
+} from 'web3-common';
+
+export default class IpcProvider<
+	API extends Web3APISpec = EthExecutionAPI,
+> extends Web3BaseProvider<API> {
+	private readonly _emitter: EventEmitter = new EventEmitter();
+
+	private readonly _socketPath: string;
+	private readonly _socket: Socket;
+	private _connectionStatus: Web3BaseProviderStatus;
+
+	private readonly _requestQueue: Map<JsonRpcId, DeferredPromise<unknown>>;
+
+	public constructor(socketPath: string) {
+		super();
+
+		this._connectionStatus = 'disconnected';
+		this._socketPath = socketPath;
+		this._socket = new Socket();
+
+		this._requestQueue = new Map<JsonRpcId, DeferredPromise<unknown>>();
+
+		this.connect();
+	}
+
+	public getStatus(): Web3BaseProviderStatus {
+		return this._connectionStatus;
+	}
+
+	/* eslint-disable class-methods-use-this */
+	public supportsSubscriptions(): boolean {
+		return true;
+	}
+
+	public on<T = JsonRpcResult>(
+		type: 'message' | 'connect' | 'disconnect' | string,
+		callback: Web3BaseProviderCallback<T>,
+	): void {
+		this._emitter.on(type, callback);
+	}
+
+	public once<T = JsonRpcResult>(type: string, callback: Web3BaseProviderCallback<T>): void {
+		this._emitter.once(type, callback);
+	}
+
+	public removeListener(type: string, callback: Web3BaseProviderCallback): void {
+		this._emitter.removeListener(type, callback);
+	}
```

Comment:
Should we be validating `callback` as done in [1.x](https://github.com/ChainSafe/web3.js/blob/1.x/packages/web3-providers-ipc/src/index.js#L225)?

Reply:
We are just passing it to the internal `this._emitter.removeListener`, if any validation required to remove it will be done `this._emitter.removeListener`